controller/appcontroller.go

```diff
@@ -1008,7 +1013,11 @@ func (ctrl *ApplicationController) processAppRefreshQueueItem() (processNext boo
 		if err := ctrl.cache.GetAppManagedResources(app.Name, &managedResources); err != nil {
 			logCtx.Warnf("Failed to get cached managed resources for tree reconciliation, fallback to full reconciliation")
 		} else {
-			if tree, err := ctrl.getResourceTree(app, managedResources); err != nil {
+			var tree *appv1.ApplicationTree
+			if err = argo.ValidateDestination(context.Background(), &app.Spec.Destination, ctrl.db); err == nil {
+				tree, err = ctrl.getResourceTree(app, managedResources)
```

Comment:
Any reason not moving https://github.com/argoproj/argo-cd/blob/a80df6f219edbcb62d4e810d074bfc674d82c732/controller/appcontroller.go#L1033 here?

Reply:
thank you for suggestion. It is actually easier to read code this way. Refactored as you suggested: https://github.com/argoproj/argo-cd/pull/4039/commits/25b15cf08ee206c67bd0cf47b0c0808372552793